### PR TITLE
Add support for non-standard Openstack networks

### PR DESCRIPTION
**Pull request by jonathanhoskin**
In my Openstack environment, I have networks with names like 'PROD', 'STAGING', etc.

I have had to modify this gem so that I can specify the bootstrap network by name. If I don't, Fog gets a 500 from the server when trying to query the API for floating IPs.

This is analogous to the `--bootstrap-network` option available in `knife openstack server create`.

In my machine options, I can then specify:

      machine_options = {
        :bootstrap_network => BOOTSTRAP_NETWORK,  << ADDED THIS
        :sudo => true,
        :ssh_username => SSH_USERNAME,
        :bootstrap_options => {
          ...
        }
      }

It works fine for me. Am I barking up the right tree?

### CODE REVIEW

**Comment by jkeiser** [comment]
Hmm, it seems reasonable given that we already have an option of the same name in knife.

**Comment by jkeiser** [comment]
@jonathanhoskin two things:

1. The bootstrap network needs to be saved into machine_location here (we try not to rely on `machine_options` which aren't always available): https://github.com/chef/chef-provisioning-fog/blob/master/lib/chef/provisioning/fog_driver/driver.rb#L334-L336 .  The reason for this is that we want to be able to run commands against the node even if we don't have the original recipe available--and we also want to protect against recipe drift (we still need to be able to get to it at the old location even if you change the networks around for new machines).

2. Is this OpenStack-specific?  It seems like it might not be given the genericity of the words.  If it isn't OpenStack-specific, then it's a fine location--but otherwise, you need to override `ssh_transport_for` in https://github.com/chef/chef-provisioning-fog/blob/master/lib/chef/provisioning/fog_driver/providers/openstack.rb to provide this functionality.

**Comment by jonathanhoskin** [comment]
@jkeiser thanks for the comments.

1. Sure, I'll figure out what machine_location does and move it there. Stand by.

2. No, it's not Openstack-specific. As I learnt just today by playing with Rackspace private networks, the same problem occurs there also.

File: lib/chef/provisioning/fog_driver/driver.rb [hunk 1]
```diff
@@ -629,7 +633,14 @@ def create_ssh_transport(machine_spec, machine_options, server)
       end
 
       remote_host = nil
-      if machine_spec.location['use_private_ip_for_ssh']
+      if machine_options.has_key?(:bootstrap_network)
```
Comment (jkeiser) [inline lib/chef/provisioning/fog_driver/driver.rb]: This spot will need machine_spec.location['bootstrap_network'] when location is fixed.

Comment (jonathanhoskin) [inline lib/chef/provisioning/fog_driver/driver.rb]: Sorry about the delay, I'm 7 weeks into an 8 week devops project. Very close to supplying a proper bit of code for this!

**Comment by jjasghar** [comment]
@jonathanhoskin can you rebase off master here? I'd be curious to see this getting in.

**Comment by jonathanhoskin** [comment]
@jjasghar sorry for the delay, this is now done!

**Comment by jjasghar** [comment]
@jonathanhoskin :metal: awesome thanks for this. I'll run it through some testing, and comment back here when I'm done.

**Comment by jjasghar** [comment]
@jonathanhoskin I'm not able to pin point the error, but my OpenStack cloud started throwing 500's when i used your patch:


```
  ================================================================================
    Error executing action `converge` on resource 'machine_batch[dev]'
    ================================================================================

    Excon::Errors::InternalServerError
    ----------------------------------
    Expected([200, 202]) <=> Actual(500 InternalServerError)
    excon.error.response
      :body          => "{\"computeFault\": {\"message\": \"The server has either erred or is incapable of performing the requested operation.\", \"code\": 500}}"
      :headers       => {
        "Content-Length"       => "128"
        "Content-Type"         => "application/json; charset=UTF-8"
        "Date"                 => "Thu, 12 Nov 2015 23:39:23 GMT"
        "X-Compute-Request-Id" => "req-5c54aae4-4935-436d-a5de-e7f7f2139b55"
      }
      :local_address => "10.0.1.8"
      :local_port    => 49216
      :reason_phrase => "Internal Server Error"
      :remote_ip     => "209.198.222.119"
      :status        => 500
      :status_line   => "HTTP/1.1 500 Internal Server Error\r\n"

    Resource Declaration:
    ---------------------
    # In /Users/jasghar/repo/chef-provisioning-fog-example/demo.rb

     14: machine_batch 'dev' do
     15:   1.upto(3) do |n|
     16:     instance = "#{name}-webserver-#{n}"
     17:     machine instance do
     18:       tag "#{name}-webserver-#{n}"
     19:       recipe 'demo-chef-provisioning::default'
     20:       converge true
     21:     end
     22:   end
     23: end

    Compiled Resource:
    ------------------
    # Declared in /Users/jasghar/repo/chef-provisioning-fog-example/demo.rb:14:in `from_file'

    machine_batch("dev") do
      action [:converge]
      retries 0
      retry_delay 2
      default_guard_interpreter :default
      driver "fog:OpenStack"
      chef_server {:chef_server_url=>"chefzero://localhost:8889", :options=>{:client_name=>"dex.local", :signing_key_filename=>nil, :api_version=>"0"}}
      machine_options {:bootstrap_options=>{:flavor_ref=>3, :image_ref=>"741b48c5-3575-4e91-a3ee-0c1f2c38432b", :nics=>[{:net_id=>"8fc162c0-311e-4006-ba81-89afddb9db6c"}], :key_name=>"jdizzle", :floating_ip_pool=>"ext-net"}, :ssh_username=>"ubuntu"}
      declared_type :machine_batch
      cookbook_name "@recipe_files"
      recipe_name "/Users/jasghar/repo/chef-provisioning-fog-example/demo.rb"
      machines ["dev-webserver-1", "dev-webserver-2", "dev-webserver-3"]
    end

[2015-11-12T17:38:48-06:00] INFO: Running queued delayed notifications before re-raising exception

Running handlers:
[2015-11-12T17:38:48-06:00] ERROR: Running exception handlers
Running handlers complete
[2015-11-12T17:38:48-06:00] ERROR: Exception handlers complete
Chef Client failed. 0 resources updated in 13 seconds
[2015-11-12T17:38:48-06:00] FATAL: Stacktrace dumped to /Users/jasghar/repo/chef-provisioning-fog-example/.chef/local-mode-cache/cache/chef-stacktrace.out
[2015-11-12T17:38:48-06:00] ERROR: machine_batch[dev] (@recipe_files::/Users/jasghar/repo/chef-provisioning-fog-example/demo.rb line 14) had an error: Excon::Errors::InternalServerError: Expected([200, 202]) <=> Actual(500 InternalServerError)
excon.error.response
  :body          => "{\"computeFault\": {\"message\": \"The server has either erred or is incapable of performing the requested operation.\", \"code\": 500}}"
  :headers       => {
    "Content-Length"       => "128"
    "Content-Type"         => "application/json; charset=UTF-8"
    "Date"                 => "Thu, 12 Nov 2015 23:39:23 GMT"
    "X-Compute-Request-Id" => "req-5c54aae4-4935-436d-a5de-e7f7f2139b55"
  }
  :local_address => "10.0.1.8"
  :local_port    => 49216
  :reason_phrase => "Internal Server Error"
  :remote_ip     => "209.198.222.119"
  :status        => 500
  :status_line   => "HTTP/1.1 500 Internal Server Error\r\n"

[2015-11-12T17:38:48-06:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
```

I rolled your patch back and it started working. :-/

**Comment by jonathanhoskin** [comment]
Sorry, I no longer have access to an Openstack cluster, so this would be hard for me to debug.

The best I could suggest is crank the logging up and find out what json is being sent in the POST.

**Comment by jjasghar** [comment]
Closing due to lack of progress.

**Pull request closed without merging**